expression/builtin_arithmetic.go

```diff
@@ -416,42 +479,44 @@ func (s *builtinArithmeticMinusIntSig) evalInt(row chunk.Row) (val int64, isNull
+	isLHSUnsigned := mysql.HasUnsignedFlag(s.args[0].GetType().Flag)
+	isRHSUnsigned := mysql.HasUnsignedFlag(s.args[1].GetType().Flag)
 
-	if forceToSigned && mysql.HasUnsignedFlag(s.args[0].GetType().Flag) {
-		if a < 0 {
-			return 0, true, types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", s.args[0].String(), s.args[1].String()))
+	switch {
+	case forceToSigned && isLHSUnsigned && isRHSUnsigned:
+		err := s.checkOverflowFUU(a, b)
+		if err != nil {
+			return 0, true, err
 		}
-	}
-	if forceToSigned && mysql.HasUnsignedFlag(s.args[1].GetType().Flag) {
-		if b < 0 {
-			return 0, true, types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", s.args[0].String(), s.args[1].String()))
+	case forceToSigned && isLHSUnsigned && !isRHSUnsigned:
+		err := s.checkOverflowFUS(a, b)
+		if err != nil {
+			return 0, true, err
 		}
-	}
-
-	switch {
-	case isLHSUnsigned && isRHSUnsigned:
-		if uint64(a) < uint64(b) {
-			return 0, true, types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", s.args[0].String(), s.args[1].String()))
+	case forceToSigned && !isLHSUnsigned && isRHSUnsigned:
+		err := s.checkOverflowFSU(a, b)
+		if err != nil {
+			return 0, true, err
 		}
-	case isLHSUnsigned && !isRHSUnsigned:
-		if b >= 0 && uint64(a) < uint64(b) {
-			return 0, true, types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", s.args[0].String(), s.args[1].String()))
+	case !isLHSUnsigned && !isRHSUnsigned:
+		err := s.checkOverflowSS(a, b)
+		if err != nil {
+			return 0, true, err
 		}
-		if b < 0 && uint64(a) > math.MaxUint64-uint64(-b) {
-			return 0, true, types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", s.args[0].String(), s.args[1].String()))
+	case !forceToSigned && isLHSUnsigned && isRHSUnsigned:
+		err := s.checkOverflowUU(a, b)
+		if err != nil {
+			return 0, true, err
 		}
-	case !isLHSUnsigned && isRHSUnsigned:
-		if a < 0 || uint64(a) < uint64(b) {
-			return 0, true, types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", s.args[0].String(), s.args[1].String()))
+	case !forceToSigned && isLHSUnsigned && !isRHSUnsigned:
+		err := s.checkOverflowUS(a, b)
+		if err != nil {
+			return 0, true, err
 		}
-	case !isLHSUnsigned && !isRHSUnsigned:
-		// We need `(a >= 0 && b == math.MinInt64)` due to `-(math.MinInt64) == math.MinInt64`.
-		// If `a<0 && b<=0`: `a-b` will not overflow even though b==math.MinInt64.
-		// If `a<0 && b>0`: `a-b` will not overflow only if `math.MinInt64<=a-b` satisfied
-		if (a >= 0 && b == math.MinInt64) || (a > 0 && -b > math.MaxInt64-a) || (a < 0 && -b < math.MinInt64-a) {
-			return 0, true, types.ErrOverflow.GenWithStackByArgs("BIGINT", fmt.Sprintf("(%s - %s)", s.args[0].String(), s.args[1].String()))
+	case !forceToSigned && !isLHSUnsigned && isRHSUnsigned:
+		err := s.checkOverflowSU(a, b)
+		if err != nil {
```

Comment:
Can we check the err outside of switch-case

Reply:
The problem is to detect overflow, we need signess of args and `forcedToSigned` flag, do you have any suggestion on how to distinguish different args and flags, without switch statement?